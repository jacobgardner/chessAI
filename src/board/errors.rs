use std::fmt::{self, Display, Formatter};


// TODO: Do we need to use failure?

#[derive(Debug, PartialEq)]
pub enum InvalidStringReason {
    IncorrectLength,
    NonAsciiChars,
}

impl Display for InvalidStringReason {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        match self {
            InvalidStringReason::IncorrectLength => write!(
                formatter,
                "Expected string to have exactly 64 non-space characters"
            ),
            InvalidStringReason::NonAsciiChars => {
                write!(formatter, "Detected 1 or more non-ascii characters")
            }
        }
    }
}

#[derive(Debug, Fail, PartialEq)]
pub enum BoardError {
    #[fail(display = "invalid player id: {}", player_id)]
    InvalidPlayer { player_id: u8 },

    #[fail(display = "invalid piece id: {}", piece_id)]
    InvalidPiece { piece_id: u8 },

    #[fail(display = "Bit found on player mask, but no board masks")]
    MalformedBoard,

    #[fail(display = "Rank/File exceeded board limits: {} {}", rank, file)]
    OutOfBounds { rank: u8, file: u8 },

    #[fail(display = "Malformed string for board: {}", _0)]
    InvalidString(InvalidStringReason),
}
