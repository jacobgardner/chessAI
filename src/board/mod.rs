use std::fmt;
use std::fmt::{Debug, Display, Formatter};

use num;

#[derive(Debug, PartialEq)]
pub enum InvalidStringReason {
    IncorrectLength,
    NonAsciiChars
}

impl Display for InvalidStringReason {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        match self {
            InvalidStringReason::IncorrectLength => write!(formatter, "Expected string to have exactly 64 non-space characters"),
            InvalidStringReason::NonAsciiChars => write!(formatter, "Detected 1 or more non-ascii characters")
        }
    }
}

#[derive(Debug, Fail, PartialEq)]
pub enum BoardError {
    #[fail(display = "invalid player id: {}", player_id)]
    InvalidPlayer { player_id: u8 },

    #[fail(display = "invalid piece id: {}", piece_id)]
    InvalidPiece { piece_id: u8 },

    #[fail(display = "Bit found on player mask, but no board masks")]
    MalformedBoard,

    #[fail(
        display = "Rank/File exceeded board limits: {} {}",
        rank,
        file
    )]
    OutOfBounds { rank: u8, file: u8 },

    #[fail(display = "Malformed string for board: {}", _0)]
    InvalidString(InvalidStringReason),
}

#[derive(Debug, PartialEq, FromPrimitive)]
pub enum Player {
    Black = 0,
    White = 1,
}

impl From<char> for Player {
    fn from(chr: char) -> Player {
        if chr.is_lowercase() {
            Player::Black
        } else {
            Player::White
        }
    }
}

#[derive(Debug, PartialEq, FromPrimitive)]
pub enum PieceType {
    Pawn = 0,
    Rook = 1,
    Knight = 2,
    Bishop = 3,
    Queen = 4,
    King = 5,
}

impl PieceType {
    fn from(chr: char) -> Option<PieceType> {
        let piece_type = match chr.to_lowercase().next().unwrap_or('x') {
            'p' => PieceType::Pawn,
            'r' => PieceType::Rook,
            'n' => PieceType::Knight,
            'b' => PieceType::Bishop,
            'q' => PieceType::Queen,
            'k' => PieceType::King,
            _ => return None,
        };

        Some(piece_type)
    }
}

#[derive(Debug, PartialEq)]
pub struct Piece {
    piece_type: PieceType,
    player: Player,
}

impl Piece {
    pub fn from(chr: char) -> Option<Piece> {
        if let Some(piece_type) = PieceType::from(chr) {
            Some(Piece {
                piece_type,
                player: Player::from(chr),
            })
        } else {
            None
        }
    }
}

const PIECE_COUNT: usize = 6;
const PLAYER_COUNT: usize = 2;
// const BOARD_COUNT: usize = PIECE_COUNT + PLAYER_COUNT;

#[derive(PartialEq)]
pub struct Board {
    pub pieces: [u64; PIECE_COUNT],
    pub players: [u64; PLAYER_COUNT],
}

struct BitPosition(u8);
impl From<(u8, u8)> for BitPosition {
    fn from((rank, file): (u8, u8)) -> BitPosition {
        BitPosition(rank * 8 + file)
    }
}

struct PositionMask(u64);

impl From<(u8, u8)> for PositionMask {
    fn from((rank, file): (u8, u8)) -> PositionMask {
        PositionMask(1u64 << BitPosition::from((rank, file)).0)
    }
}

impl Board {
    pub fn piece_at(&self, rank: u8, file: u8) -> Result<Option<Piece>, BoardError> {
        if rank >= 8 || file >= 8 {
            return Err(BoardError::OutOfBounds { rank, file });
        }

        let mask = PositionMask::from((rank, file)).0;

        if let Some((player_id, _)) = self
            .players
            .iter()
            .enumerate()
            .find(|&(_, player_board)| mask & player_board > 0)
        {
            let player =
                num::FromPrimitive::from_usize(player_id).ok_or(BoardError::InvalidPlayer {
                    player_id: player_id as u8,
                })?;

            let (i, _) = self
                .pieces
                .iter()
                .enumerate()
                .find(|&(_, board)| mask & board > 0)
                .ok_or(BoardError::MalformedBoard)?;

            debug_assert!(i < PIECE_COUNT);

            let piece_type = num::FromPrimitive::from_usize(i)
                .ok_or(BoardError::InvalidPiece { piece_id: i as u8 })?;

            // let piece_board = (0..PIECE_COUNT).find(|&i| mask & self.pieces[i] > 0).ok_or(())?;
            Ok(Some(Piece { player, piece_type }))
        } else {
            debug_assert!({
                (0..PIECE_COUNT)
                    .find(|&i| mask & self.pieces[i] > 0)
                    .is_none()
            });

            Ok(None)
        }
    }

    pub fn from(board: &str) -> Result<Board, BoardError> {
        let mut pieces = [0; PIECE_COUNT];
        let mut players = [0; PLAYER_COUNT];

        let board: String = board.split(char::is_whitespace).collect();

        if board.len() != 64 {
            return Err(BoardError::InvalidString(InvalidStringReason::IncorrectLength));
        }

        // TODO: Make sure this correctly throws an error on non-ascii
        for (i, chr) in board.chars().enumerate() {
            if !chr.is_ascii() {
                return Err(BoardError::InvalidString(InvalidStringReason::NonAsciiChars));
            }

            let rank: u8 = (7 - (i / 8)) as u8;
            let file: u8 = (i % 8) as u8;

            let piece_mask = PositionMask::from((rank, file)).0;

            if let Some(piece) = Piece::from(chr) {
                players[piece.player as usize] |= piece_mask;
                pieces[piece.piece_type as usize] |= piece_mask;
            }
        }

        Ok(Board { pieces, players })
    }
}

impl Debug for Board {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        Display::fmt(self, formatter)
    }
}

impl Display for Board {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        let mut board = String::with_capacity(128);

        board += "       +--------+\n";

        for r in 0..8 {
            // let rank_chr = (65u8 + (7 - r as u8))  as char;

            board += &format!("0x{: <02x} {} |", (7 - r) * 8, 8 - r);

            for f in 0..8 {
                let piece = self.piece_at(7 - r, f).map_err(|_| fmt::Error)?;

                // let piece = Some(Piece {
                //     piece_type: PieceType::Pawn,
                //     player: Player::White
                // });

                let chr = if let Some(piece) = piece {
                    let piece_chr = match piece.piece_type {
                        PieceType::Pawn => 'p',
                        PieceType::Rook => 'r',
                        PieceType::Bishop => 'b',
                        PieceType::Knight => 'n',
                        PieceType::King => 'k',
                        PieceType::Queen => 'q',
                    };

                    if piece.player == Player::White {
                        piece_chr.to_ascii_uppercase()
                    } else {
                        piece_chr
                    }
                } else {
                    '.'
                };

                board += &chr.to_string();
            }
            board += &format!("| {}\n", (8 - r) * 8 - 1);
        }

        board += "       +--------+\n";
        board += "        ABCDEFGH\n";

        write!(formatter, "{}", board)
    }
}

#[test]
fn test_board_from_str() {
    let board = Board::from(
        "
    xxxxxxxx
    xxxxxxxx
    xxxxxxxx
    xxxxxxxx
    xxxxxxxx
    xxxxxxxx
    xxxxxxxx
    xxxxxxxx
    ",
    ).unwrap();

    assert_eq!(Board::from(""), Err(BoardError::InvalidString(InvalidStringReason::IncorrectLength)));
    assert_eq!(board.players[0] | board.players[1], 0);

    let board = Board::from(
        "
    .......r
    ...P....
    ........
    ........
    ..k.....
    ........
    .Q...P..
    ..P.....
    ",
    ).unwrap();

    let rook_mask = PositionMask::from((7, 7)).0;
    let king_mask = PositionMask::from((3, 2)).0;

    let queen_mask = PositionMask::from((1, 1)).0;
    let pawn_mask =
        PositionMask::from((0, 2)).0 | PositionMask::from((1, 5)).0 | PositionMask::from((6, 3)).0;

    let black_mask = rook_mask | king_mask;
    let white_mask = queen_mask | pawn_mask;

    assert_eq!(board.players[Player::Black as usize], black_mask);
    assert_eq!(board.players[Player::White as usize], white_mask);

    assert_eq!(board.pieces[PieceType::Pawn as usize], pawn_mask);
    assert_eq!(board.pieces[PieceType::Queen as usize], queen_mask);
    assert_eq!(board.pieces[PieceType::Rook as usize], rook_mask);
    assert_eq!(board.pieces[PieceType::King as usize], king_mask);

    let board = Board::from(
        "
    .......r
    ...ðŸ˜€...
    ........
    ......
    ..k.....
    ........
    .Q...P..
    ..P.....
    ",
    );

    assert_eq!(board, Err(BoardError::InvalidString(InvalidStringReason::NonAsciiChars)));
}

#[test]
fn test_piece_at() {
    let pieces: [u64; PIECE_COUNT] = [1, 1 << 8, 1 << 12, 1 << 16, 1 << 25, 1 << 63];

    let board = Board {
        players: [
            pieces[0] | pieces[2] | pieces[4] | 1 << 5,
            pieces[1] | pieces[3] | pieces[5],
        ],
        pieces: pieces,
    };

    assert_eq!(
        board.piece_at(0, 0).unwrap(),
        Some(Piece {
            piece_type: PieceType::Pawn,
            player: Player::Black
        })
    );

    assert_eq!(
        board.piece_at(1, 0).unwrap(),
        Some(Piece {
            piece_type: PieceType::Rook,
            player: Player::White
        })
    );

    assert_eq!(
        board.piece_at(1, 4).unwrap(),
        Some(Piece {
            piece_type: PieceType::Knight,
            player: Player::Black
        })
    );

    assert_eq!(
        board.piece_at(2, 0).unwrap(),
        Some(Piece {
            piece_type: PieceType::Bishop,
            player: Player::White
        })
    );

    assert_eq!(
        board.piece_at(3, 1).unwrap(),
        Some(Piece {
            piece_type: PieceType::Queen,
            player: Player::Black
        })
    );

    assert_eq!(
        board.piece_at(7, 7).unwrap(),
        Some(Piece {
            piece_type: PieceType::King,
            player: Player::White
        })
    );

    assert_eq!(board.piece_at(4, 4).unwrap(), None);
    assert_eq!(board.piece_at(0, 5), Err(BoardError::MalformedBoard));
}

#[test]
fn test_player_from_str() {
    assert_eq!(Player::from(&'Q'), Player::White);
    assert_eq!(Player::from(&'K'), Player::White);
    assert_eq!(Player::from(&'N'), Player::White);
    assert_eq!(Player::from(&'B'), Player::White);
    assert_eq!(Player::from(&'P'), Player::White);
    assert_eq!(Player::from(&'R'), Player::White);
    assert_eq!(Player::from(&'X'), Player::White);

    assert_eq!(Player::from(&'q'), Player::Black);
    assert_eq!(Player::from(&'k'), Player::Black);
    assert_eq!(Player::from(&'n'), Player::Black);
    assert_eq!(Player::from(&'b'), Player::Black);
    assert_eq!(Player::from(&'p'), Player::Black);
    assert_eq!(Player::from(&'r'), Player::Black);
    assert_eq!(Player::from(&'x'), Player::Black);
}

#[test]
fn test_to_piece_owner() {
    assert_eq!(
        Piece::from(&'P'),
        Some(Piece {
            piece_type: PieceType::Pawn,
            player: Player::White
        })
    );
    assert_eq!(
        Piece::from(&'p'),
        Some(Piece {
            piece_type: PieceType::Pawn,
            player: Player::Black
        })
    );

    assert_eq!(
        Piece::from(&'K'),
        Some(Piece {
            piece_type: PieceType::King,
            player: Player::White
        })
    );
    assert_eq!(
        Piece::from(&'k'),
        Some(Piece {
            piece_type: PieceType::King,
            player: Player::Black
        })
    );

    assert_eq!(Piece::from(&'l'), None);
}

#[test]
fn test_piece_from_str() {
    assert_eq!(PieceType::from(&'P'), Some(PieceType::Pawn));
    assert_eq!(PieceType::from(&'p'), Some(PieceType::Pawn));

    assert_eq!(PieceType::from(&'R'), Some(PieceType::Rook));
    assert_eq!(PieceType::from(&'r'), Some(PieceType::Rook));

    assert_eq!(PieceType::from(&'N'), Some(PieceType::Knight));
    assert_eq!(PieceType::from(&'n'), Some(PieceType::Knight));

    assert_eq!(PieceType::from(&'B'), Some(PieceType::Bishop));
    assert_eq!(PieceType::from(&'b'), Some(PieceType::Bishop));

    assert_eq!(PieceType::from(&'Q'), Some(PieceType::Queen));
    assert_eq!(PieceType::from(&'q'), Some(PieceType::Queen));

    assert_eq!(PieceType::from(&'K'), Some(PieceType::King));
    assert_eq!(PieceType::from(&'k'), Some(PieceType::King));

    assert_eq!(PieceType::from(&'x'), None);
    assert_eq!(PieceType::from(&'L'), None);
}
