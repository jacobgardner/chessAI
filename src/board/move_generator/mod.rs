mod pawn;

use super::Board;
use super::{PieceType, Player, PIECE_COUNT};

pub struct MoveGenerator {
    root_board: Board,
    player: Player,

    player_mask: u64,
    enemy_mask: u64,
    all_pieces: u64,
    player_piecetype_mask: u64,

    is_first_move: bool,
    available_moves: u64,
    available_captures: u64,

    piece_index: usize,
}

impl MoveGenerator {
    pub fn new(root_board: Board, player: Player) -> Self {
        let player_mask = root_board.players[player as usize];
        let enemy_mask = root_board.players[1 - (player as usize)];
        let all_pieces = player_mask | enemy_mask;

        let mut gen = MoveGenerator {
            root_board,
            player,

            player_mask,
            enemy_mask,
            all_pieces,

            player_piecetype_mask: 0,

            is_first_move: true,
            available_moves: 0,
            available_captures: 0,

            piece_index: 0,
        };

        gen.player_piecetype_mask = gen.generate_piecemask(0);

        gen
    }

    fn generate_piecemask(&self, piece_index: usize) -> u64 {
        self.root_board.pieces[piece_index as usize] & self.player_mask
    }
}

impl Iterator for MoveGenerator {
    type Item = Board;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if self.player_piecetype_mask == 0 {
                self.piece_index += 1;
                self.is_first_move = true;

                if self.piece_index < PIECE_COUNT {
                    self.player_piecetype_mask = self.generate_piecemask(self.piece_index);
                    // Restarting the loop because we can't be sure the next mask > 0
                    continue;
                } else {
                    return None;
                }
            }

            assert!(self.player_piecetype_mask > 0);
            assert!(self.piece_index < PIECE_COUNT);

            // TODO: A lot of these can be cached
            let index = self.player_piecetype_mask.trailing_zeros();
            let piece_mask = 1 << index;
            let piece_inverse = !piece_mask;

            let piece_type: PieceType = num::FromPrimitive::from_usize(self.piece_index).unwrap();

            match piece_type {
                PieceType::Pawn => match self.generate_next_pawn_move(index, piece_mask) {
                    Some(board) => {
                        return Some(board);
                    }
                    None => {
                        self.is_first_move = true;
                        self.player_piecetype_mask &= piece_inverse;
                    }
                },
                _ => {
                    // TODO: We'll want to remove the piece from the mask if there are no
                    //  moves left.
                    self.player_piecetype_mask &= piece_inverse;
                }
            }
        }
    }
}
