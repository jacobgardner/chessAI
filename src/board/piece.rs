use super::piece_type::PieceType;
use super::player::Player;

#[derive(Debug, PartialEq)]
pub struct Piece {
    pub(super) piece_type: PieceType,
    pub(super) player: Player,
}

impl Piece {
    pub fn from(chr: char) -> Option<Piece> {
        if let Some(piece_type) = PieceType::from(chr) {
            Some(Piece {
                piece_type,
                player: Player::from(chr),
            })
        } else {
            None
        }
    }
}

#[test]
fn test_to_piece_owner() {
    assert_eq!(
        Piece::from('P'),
        Some(Piece {
            piece_type: PieceType::Pawn,
            player: Player::White
        })
    );
    assert_eq!(
        Piece::from('p'),
        Some(Piece {
            piece_type: PieceType::Pawn,
            player: Player::Black
        })
    );

    assert_eq!(
        Piece::from('K'),
        Some(Piece {
            piece_type: PieceType::King,
            player: Player::White
        })
    );
    assert_eq!(
        Piece::from('k'),
        Some(Piece {
            piece_type: PieceType::King,
            player: Player::Black
        })
    );

    assert_eq!(Piece::from('l'), None);
}
