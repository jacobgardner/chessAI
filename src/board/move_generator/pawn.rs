use super::MoveGenerator;
use crate::bitboard::BitBoard;
use crate::bitboard::{ROW_1, ROW_2, ROW_7, ROW_8};
use crate::bitposition::BitPosition;
use crate::board::{Board, PieceType, Player, PIECE_COUNT};

// TODO: En Passant (https://en.wikipedia.org/wiki/En_passant)

impl MoveGenerator {
    // TODO: We may be able to make this mostly work for other pieces as well
    fn move_pawn(
        &mut self,
        current_position_mask: BitBoard,
        next_position_mask: BitBoard,
        capture: bool,
    ) -> Board {
        let mut board = self.root_board.clone();

        // Remove current position from pawn and current player bitboards
        board.pieces[PieceType::Pawn as usize] =
            board.pieces[PieceType::Pawn as usize].intersect(current_position_mask.inverse());
        board.players[self.player as usize] =
            board.players[self.player as usize].intersect(current_position_mask.inverse());

        if capture {
            debug_assert!(
                board.players[self.player as usize]
                    .intersect(next_position_mask)
                    .is_empty(),
                "Pawn Move Invariant Invalidated: Capture move made on space occupied by self"
            );

            debug_assert!(
                !board.players[1 - (self.player as usize)].intersect(next_position_mask).is_empty(),
                "Pawn Move Invariant Invalidated: Capture move made on space not-occupied by opponent"
            );

            // Because this is a capture we need to remove the previous piece
            for i in 0..PIECE_COUNT {
                board.pieces[i] = board.pieces[i].intersect(next_position_mask.inverse());
            }

            // And the previous player
            board.players[1 - (self.player as usize)] =
                board.players[1 - (self.player as usize)].intersect(next_position_mask.inverse());
        } else {
            debug_assert!(
                board.players[self.player as usize]
                    .intersect(next_position_mask)
                    .is_empty(),
                "Pawn Move Invariant Invalidated: Non-capture move made on space occupied by self"
            );
            debug_assert!(
                board.players[1 - (self.player as usize)].intersect(next_position_mask).is_empty(),
                "Pawn Move Invariant Invalidated: Non-capture move made on space occupied by opponent"
            );
        }

        board.players[self.player as usize] =
            board.players[self.player as usize].join(next_position_mask);

        if !next_position_mask.intersect(ROW_1).is_empty()
            || !next_position_mask.intersect(ROW_8).is_empty()
        {
            board.pieces[PieceType::Queen as usize] =
                board.pieces[PieceType::Queen as usize].join(next_position_mask);
        } else {
            board.pieces[PieceType::Pawn as usize] =
                board.pieces[PieceType::Pawn as usize].join(next_position_mask);
        }

        board
    }

    // TODO: index and current_position_mask represent the same thing.  Do we need both?
    #[inline(always)]
    fn pawn_available_moves(
        &self,
        index: BitPosition,
        current_position_mask: BitBoard,
    ) -> BitBoard {
        debug_assert!(
            current_position_mask
                .intersect(ROW_1.join(ROW_8))
                .is_empty(),
            "Pawn Invariant Invalidation: Pawn must never appear in the first or last row"
        );

        // TODO: Our snapshot tests caught this bug occuring but it wasn't able to pinpoint
        //  where the bug was occuring.  We need more granular unit tests to catch stuff
        //  like this.
        let mut moves = self.all_pieces.inverse().intersect(match self.player {
            Player::White => BitBoard::from(index.shift(0, 1)),
            Player::Black => BitBoard::from(index.shift(0, -1)),
        });

        if !moves.is_empty() {
            moves = moves.join(self.all_pieces.inverse().intersect(match self.player {
                Player::White => {
                    if !current_position_mask.intersect(ROW_2).is_empty() {
                        BitBoard::from(index.shift(0, 2))
                    } else {
                        BitBoard::empty()
                    }
                }
                Player::Black => {
                    if !current_position_mask.intersect(ROW_7).is_empty() {
                        BitBoard::from(index.shift(0, -2))
                    } else {
                        BitBoard::empty()
                    }
                }
            }))
        }

        moves
    }

    #[inline(always)]
    fn pawn_captures(&self, index: BitPosition) -> BitBoard {
        self.enemy_mask.intersect(match self.player {
            Player::White => if !index.is_leftmost() {
                BitBoard::from(index.shift(-1, 1))
            } else {
                BitBoard::empty()
            }
            .join(if !index.is_rightmost() {
                BitBoard::from(index.shift(1, 1))
            } else {
                BitBoard::empty()
            }),
            Player::Black => if !index.is_leftmost() {
                BitBoard::from(index.shift(-1, -1))
            } else {
                BitBoard::empty()
            }
            .join(if !index.is_rightmost() {
                BitBoard::from(index.shift(1, -1))
            } else {
                BitBoard::empty()
            }),
        })
    }

    pub(crate) fn generate_next_pawn_move(
        &mut self,
        index: BitPosition,
        current_position_mask: BitBoard,
    ) -> Option<Board> {
        if self.is_first_move {
            self.available_moves = self.pawn_available_moves(index, current_position_mask);
            self.available_captures = self.pawn_captures(index);

            self.is_first_move = false;
        }

        if !self.available_moves.is_empty() {
            // We can probably join these two
            let new_move = self.available_moves.right_position();
            let next_position_mask = BitBoard::from(new_move);

            let board = self.move_pawn(current_position_mask, next_position_mask, false);

            self.available_moves = self.available_moves.intersect(next_position_mask.inverse());

            return Some(board);
        }

        if !self.available_captures.is_empty() {
            let new_move = self.available_captures.right_position();
            let next_position_mask = BitBoard::from(new_move);
            let inverse_move = next_position_mask.inverse();

            let board = self.move_pawn(current_position_mask, next_position_mask, true);

            self.available_captures = self.available_captures.intersect(inverse_move);

            return Some(board);
        }

        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use snapshot::snapshot;

    const WHITE_PAWN_TEST: &'static str = "
    xxxrxxxx
    xxPxxxxx
    xxxxPxxx
    xxxxxxxx
    xnxnxxxx
    nxPxxxxn
    xPxxxPxP
    xxxxxxxx
    ";

    const BLACK_PAWN_TEST: &'static str = "
    xxxxxxxx
    pxxxxxpx
    NxxxxnxN
    xxpxxxxx
    xxxxxxxx
    xxxxpxxx
    xxxxpNxx
    xxxNxxxx
";

    #[snapshot]
    fn test_generate_white_pawn_moves() -> Vec<String> {
        let mut boards = vec![];

        let board = Board::from(WHITE_PAWN_TEST).unwrap();
        boards.push(format!("{}", board).to_owned());

        let mut generator = board.generate_moves(Player::White);

        loop {
            let new_board = match generator.next() {
                Some(board) => board,
                None => break,
            };

            boards.push(format!("{}", new_board).to_owned());
        }
        boards
    }

    #[snapshot]
    fn test_generate_black_pawn_moves() -> Vec<String> {
        let mut boards = vec![];

        let board = Board::from(BLACK_PAWN_TEST).unwrap();
        boards.push(format!("{}", board).to_owned());

        let mut generator = board.generate_moves(Player::Black);

        loop {
            let new_board = match generator.next() {
                Some(board) => board,
                None => break,
            };

            boards.push(format!("{}", new_board).to_owned());
        }

        boards
    }
}
