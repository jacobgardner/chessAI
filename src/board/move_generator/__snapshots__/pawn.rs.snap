{
  "chess::board::move_generator::pawn::tests::test_generate_black_pawn_moves":
  {
    file:
    [
      src
      board
      move_generator
      pawn.rs
    ]
    module_path: chess::board::move_generator::pawn::tests
    test_function: test_generate_black_pawn_moves
    recorded_value:
    [
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.....p.| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.p.....| 39
        0x18 4 |...Pp...| 31
        0x10 3 |........| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.....p.| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.p.....| 39
        0x18 4 |...Pp...| 31
        0x10 3 |........| 23
        0x08 2 |.....N..| 15
        0x00 1 |...Nq...| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.....p.| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.p.....| 39
        0x18 4 |...Pp...| 31
        0x10 3 |........| 23
        0x08 2 |.....N..| 15
        0x00 1 |...q....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.....p.| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.p.....| 39
        0x18 4 |........| 31
        0x10 3 |...p....| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.....p.| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.p.....| 39
        0x18 4 |...P....| 31
        0x10 3 |....p...| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.....p.| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.......| 39
        0x18 4 |..pPp...| 31
        0x10 3 |........| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.....p.| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.......| 39
        0x18 4 |...pp...| 31
        0x10 3 |........| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.......| 55
        0x28 6 |N....N.N| 47
        0x20 5 |N.p...p.| 39
        0x18 4 |...Pp...| 31
        0x10 3 |........| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.......| 55
        0x28 6 |N....NpN| 47
        0x20 5 |N.p.....| 39
        0x18 4 |...Pp...| 31
        0x10 3 |........| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.......| 55
        0x28 6 |N....p.N| 47
        0x20 5 |N.p.....| 39
        0x18 4 |...Pp...| 31
        0x10 3 |........| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |........| 63
        0x30 7 |p.......| 55
        0x28 6 |N....N.p| 47
        0x20 5 |N.p.....| 39
        0x18 4 |...Pp...| 31
        0x10 3 |........| 23
        0x08 2 |....pN..| 15
        0x00 1 |...N....| 7
               +--------+
                ABCDEFGH

        '''
    ]
  }
  "chess::board::move_generator::pawn::tests::test_generate_white_pawn_moves":
  {
    file:
    [
      src
      board
      move_generator
      pawn.rs
    ]
    module_path: chess::board::move_generator::pawn::tests
    test_function: test_generate_white_pawn_moves
    recorded_value:
    [
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.n....| 31
        0x10 3 |n.P....n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.n....| 31
        0x10 3 |nPP....n| 23
        0x08 2 |.....P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.n....| 31
        0x10 3 |P.P....n| 23
        0x08 2 |.....P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.n....| 31
        0x10 3 |n.P..P.n| 23
        0x08 2 |.P.....P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.n.P..| 31
        0x10 3 |n.P....n| 23
        0x08 2 |.P.....P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.nPn....| 31
        0x10 3 |n......n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.P.n....| 31
        0x10 3 |n......n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.P....| 31
        0x10 3 |n......n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |.....P..| 47
        0x20 5 |........| 39
        0x18 4 |.n.n....| 31
        0x10 3 |n.P....n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...r....| 63
        0x30 7 |..P.....| 55
        0x28 6 |....P...| 47
        0x20 5 |.....p..| 39
        0x18 4 |.n.n....| 31
        0x10 3 |n.P....n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |..Qr....| 63
        0x30 7 |........| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.n....| 31
        0x10 3 |n.P....n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
      
        '''
               +--------+
        0x38 8 |...Q....| 63
        0x30 7 |........| 55
        0x28 6 |........| 47
        0x20 5 |....Pp..| 39
        0x18 4 |.n.n....| 31
        0x10 3 |n.P....n| 23
        0x08 2 |.P...P.P| 15
        0x00 1 |........| 7
               +--------+
                ABCDEFGH

        '''
    ]
  }
}